src/main/java/com/google/cloud/flink/bigquery/BigQueryDirectDataWriterContext.java

```diff
@@ -0,0 +1,162 @@
+  private BigQueryDirectDataWriterHelper writerHelper;
+  private TableId tableId;
+  private BigQueryClientFactory bigQueryWriteClientFactory;
+
+  private Optional<String> traceId;
+
+  public BigQueryDirectDataWriterContext(
+      String[] fieldNames,
+      DataType[] fieldDataTypes,
+      FlinkBigQueryConfig bqConfig,
+      BigQueryClientFactory bigQueryWriteClientFactory)
+      throws JSQLParserException {
+    this.tableId = bqConfig.getTableId();
+    this.traceId = bqConfig.getTraceId();
+    this.bigQueryWriteClientFactory = bigQueryWriteClientFactory;
+    List<DataType> columnDataTypeList = Arrays.asList(fieldDataTypes);
+    List<String> columnNameList = Arrays.asList(fieldNames);
+    ArrayList<RowField> listOfRowFields = new ArrayList<RowField>();
+    for (int i = 0; i < columnNameList.size(); i++) {
+      listOfRowFields.add(
+          new RowField(
+              columnNameList.get(i).toString(), columnDataTypeList.get(i).getLogicalType()));
+    }
+    this.flinkSchema = new RowType(listOfRowFields);
+    this.tablePath =
+        String.format(
+            "projects/%s/datasets/%s/tables/%s",
+            tableId.getProject(), tableId.getDataset(), tableId.getTable());
+
+    try {
+      this.schemaDescriptor = toDescriptor(flinkSchema);
+    } catch (Descriptors.DescriptorValidationException e) {
+      throw new BigQueryConnectorException.InvalidSchemaException(
+          "Could not convert flink-schema to descriptor object", e);
+    }
+    initializeWriterHelper();
+  }
+
+  private void initializeWriterHelper() {
+    ProtoSchema protoSchema = ProtobufUtils.toProtoSchema(flinkSchema);
+    RetrySettings.Builder retrySettingsBuilder = RetrySettings.newBuilder();
+    retrySettingsBuilder.setMaxAttempts(3);
+    RetrySettings bigqueryDataWriterHelperRetrySettings = retrySettingsBuilder.build();
+    writerHelper =
+        new BigQueryDirectDataWriterHelper(
+            bigQueryWriteClientFactory,
+            tablePath,
+            protoSchema,
+            bigqueryDataWriterHelperRetrySettings,
+            traceId);
+  }
+
+  @Override
+  public void write(Row record) throws IOException {
+    ByteString message =
+        buildSingleRowMessage(flinkSchema, schemaDescriptor, record).toByteString();
+    this.writerHelper.addRow(message);
+  }
+
+  public String getWriterStream() {
+
+    return writerHelper.getWriteStreamName();
+  }
+
+  @Override
+  public WriterCommitMessageContext finalizeStream() throws IOException {
+
+    logger.debug("Data Writer commit()");
+
+    long rowCount = writerHelper.commit();
+    String writeStreamName = writerHelper.getWriteStreamName();
+    logger.debug("Data Writer write-stream has finalized with row count: {}", rowCount);
+
+    return new BigQueryDirectWriterCommitMessageContext(writeStreamName, tablePath, rowCount);
+  }
+
+  @Override
+  public void commit() throws IOException {
+    String writeStreamName = getWriterStream();
```

Comment:
This functionality belongs in the job manager, and should take all the streams - not just one